### PR TITLE
In assisted decoding, pass model_kwargs to model's forward call (fix prepare_input_for_generation in all models)

### LINKED ISSUE #25020: GenerationMixin: model_kwargs not passed to model in assisted decoding
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
- Python version: 3.11.4
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- Accelerate version: 0.21.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@gante 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model = AutoModelForCausalLM.from_pretrained("gpt2")
assist = AutoModelForCausalLM.from_pretrained("distilgpt2")
tokenizer = AutoTokenizer.from_pretrained("gpt2")

inputs = tokenizer("The first rule of fight", return_tensors='pt')

outputs = model.generate(**inputs, token_type_ids=torch.tensor([[0,0,0,0,0]], dtype=torch.long))
print(tokenizer.decode(outputs[0]))

# output: The first rule of fight!!!!!!!!!!!!!!!

outputs = model.generate(**inputs, token_type_ids=torch.tensor([[0,0,0,0,0]], dtype=torch.long), num_beams=1, assistant_model=assist)
print(tokenizer.decode(outputs[0]))

# output: The first rule of fight-or-flight is to be prepared for the enemy. If you are
```

### Expected behavior

I would expect the outputs to be the same for the assisted generation as for the regular generation, as the token_type_ids is being passed into generate in both cases. It is expected that the `generate` method passes extra arguments to the model via its `prepare_inputs_for_generation` method.

In fact, the assisted generation does not forward the `model_kwargs` to the model as the other generation methods do.

### PR DESCRIPTION
**Pull request by sinking-point**
# What does this PR do?

Previously, assisted decoding would ignore any additional kwargs that it doesn't explicitly handle. This was inconsistent with other generation methods, which pass the model_kwargs through prepare_inputs_for_generation and forward the returned dict to the model's forward call.

The prepare_inputs_for_generation method needs to be amended in all models, as previously it only kept the last input ID when a past_key_values was passed.

Fixes #25020

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@gante

### CODE REVIEW

**Comment by gante** [comment]
Not urgent -- simply double-checking whether it was in need of a review or not :)

**Comment by sinking-point** [comment]
@gante This should be ready for review now. Thanks in advance.

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1300]:
```suggestion
    def _extend_attention_mask(self, model_kwargs: Dict[str, Any], new_mask_length: int) -> Dict[str, Any]:
```
`to` is a common PyTorch operator, so we should avoid using it as a variable (hurts readability). Feel free to pick a different variable name, as long as it is descriptive :)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1310]:
```suggestion
        mask_extension_length = new_mask_length - mask.shape[1]
```
nit: `amount` is a bit vague. We prefer long, well-defined variable names, as they facilitate maintenance 
(This change needs to be propagated below)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4457]: A `deepcopy` here will duplicate the KV cache, which increases the GPU ram requirements -- we should avoid it unless strictly needed. 

I think it would be fine to modify `model_kwargs` itself -- is there a reason why  you added the deepcopy?

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4459]: Could this one be factorized into a function as well? :)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4468]: 👌 👌 👌

File: src/transformers/models/bark/modeling_bark.py
Comment (gante) [inline src/transformers/models/bark/modeling_bark.py:497]:
```suggestion
            length_without_past = seq_len - past_length
            input_ids = input_ids[:,-length_without_past:]
```
If I'm not mistaken, this is equivalent to what you wrote -- with the added benefit of being shorter and easier to read. Would you be able to confirm? 🤗

**Review by gante — COMMENTED** [review]
@sinking-point this is a great piece of work, thank you so much for working on this 🙏 

For me, it's a green light on the logic of the PR 🟢 I've added a few comments to improve on the readability of the changes (and one performance-related comment), so that we can fulfill our role as maintainers on top of your work at super-human levels 🤗

**Comment by gante** [comment]
@ArthurZucker @LysandreJik This PR is a big one and touches in a core piece of logic for all generative models, so I'm tagging 2 core maintainers.


### Context
Advanced generation techniques (like assisted generation or [medusa](https://github.com/FasterDecoding/Medusa)) may generate more than one token per model forward pass. The original implementation of assisted generation had a lot of custom code, as this break one of the assumptions in `prepare_inputs_for_generation` -- that only one token is generated per `forward` pass.

### Solution
@sinking-point has kindly put forward a proposal to get rid of the custom code in assisted generation. After iterating with me, the plan was to remove the assumption of one token per `forward` in `prepare_inputs_for_generation` -- the slicing therein should be done based on how many tokens do not have corresponding past KV values, and not simply taking the last set of inputs. This is the change this PR implements, as well as the removal of some custom assisted generation code.

### Postface
To reiterate --  this PR gets the green light from me in terms of logic 🟢, and it is a big contribution by @sinking-point. I'll be off the next few weeks, but I'm sure this PR will get a quick resolution 🤗

**Comment by sinking-point** [comment]
Thanks @gante , I'll take a look at your comments tomorrow 👍

File: src/transformers/generation/utils.py
Comment (sinking-point) [reply src/transformers/generation/utils.py:4457]: This isn't the only thing messing with the `attention_mask` and `token_type_ids`. These are also touched by `_update_model_kwargs_for_generation`. I did the deepcopy so I don't mess with that logic. I can see how it might impact performance though.

If I naively change it to simply update model_kwargs, the `attention_mask` (presumably also the `token_type_ids`) becomes too long for the sequence length.

I could change it so that it:
1. Stores `old_attention_mask` and `old_token_type_ids`
2. Overwrites `model_kwargs` with the extended versions
3. Does the inference step
4. Restores the old values to `model_kwargs`

That should solve the performance issue. What do you think?

File: src/transformers/models/bark/modeling_bark.py
Comment (sinking-point) [reply src/transformers/models/bark/modeling_bark.py:497]: There are some generation methods (e.g. contrastive search iirc) that already only pass the last input ID. The conditional is necessary to default to the old behaviour of keeping only the last ID in such cases.

Your alternative would fail in this case. Say past is length 10.

length_without_past = 1 - 10 = -9

input_ids is length 1, so input_ids[:,9:] is out of bounds.

**Comment by Rocketknight1** [comment]
Hi @sinking-point! Sorry for the delay - I'm taking over this PR from @gante because he's out on a well-deserved rest right now. Is everything ready for review, or are there any other issues you want to discuss with the team before we take a final look at it?

**Comment by sinking-point** [comment]
No worries @Rocketknight1 . Thanks for taking this on.

There's one discussion gante opened that I haven't resolved. Could you give your input on this? https://github.com/huggingface/transformers/pull/25242#discussion_r1326154797

File: src/transformers/generation/utils.py
Comment (Rocketknight1) [reply src/transformers/generation/utils.py:4472]: Hey! I've looked at the code and I can't find any locations where the tensors in `candidate_kwargs` are mutated in place, though I might have missed something!

If that's the case, could you try replacing `copy.deepcopy`, which will copy all the underlying tensors and increase time + memory usage, with just `candidate_kwargs = model_kwargs.copy()`? This will create a shallow copy of the dict, where each of the dict values will refer to the original tensors without any actual GPU copies.

**Comment by Rocketknight1** [comment]
@sinking-point Replied to the last remaining issue up above!

**Comment by sinking-point** [comment]
Thanks @Rocketknight1 , I'll take a look on Monday

File: src/transformers/generation/utils.py
Comment (sinking-point) [reply src/transformers/generation/utils.py:4472]: Thanks, I think you're right.

**Comment by sinking-point** [comment]
Some random tests started failing so I rebased onto main where they're fixed, but it looks like I have some more work to do now.

**Comment by Rocketknight1** [comment]
Ick, yeah. I'm not sure what's causing those test failures, but if you can't figure it out, let me know and I'll dive in!

**Comment by sinking-point** [comment]
@Rocketknight1 I think I've sorted it out. The hub test fails seem unrelated to my changes, and might have been a network failure or something. The failing Musicgen pipeline is broken on main as well.

**Comment by sinking-point** [comment]
Could you block new generative model merges until this PR is merged? Otherwise I will have to keep applying the changes to any new models that are added.

**Comment by Rocketknight1** [comment]
That's a big ask, unfortunately, but they shouldn't be too frequent! Let's just try to get this PR over the line quickly, so you don't have to keep rebasing and updating.

**Comment by sinking-point** [comment]
Should be ready to merge if you're happy with it. Thanks!

**Comment by Rocketknight1** [comment]
Understood! It's quite a big PR since it touches so many models, but I'll try to get an internal review in the next few days.

**Comment by sinking-point** [comment]
> Could you block new generative model merges until this PR is merged?

Alternatively, could you require that new generative models' `prepare_inputs_for_generation` method follows this PR? That is, instead assuming that if `past_key_values` is provided it covers all but the last position, you should calculate how many positions are remaining after `past_key_values` and keep those.

**Review by gante — APPROVED** [review]
LGTM, thank you for iterating 🙏

**Comment by gante** [comment]
@sinking-point regarding the failing test: rebasing the PR should fix it, the bug was fixed last week :)

**Comment by sinking-point** [comment]
Thanks @gante :)

**Comment by LysandreJik** [comment]
This seems ok to me but I'd like to ask @patrickvonplaten for his opinion and eventual approval given the experience maintaining this part of the code

**Review by patrickvonplaten — APPROVED** [review]
Very nice! Code is cleaned up and made more extendable - very much in favor of this change

**Comment by sinking-point** [comment]
Amazing, thank you @LysandreJik and @patrickvonplaten

**Comment by gante** [comment]
And thank you @sinking-point for this big contribution 🔥 💛

**Pull request merged**
